test/SlowTests/Issues/RavenDB_11632.cs

```diff
@@ -0,0 +1,98 @@
+using FastTests;
+using Xunit;
+
+namespace SlowTests.Issues
+{
+    public class RavenDB_11632 : RavenTestBase
+    {
+        [Fact]
+        public void ShouldBeAbleToStoreExistingDocumentWithEarlierIdAfterOldDocumentWasDeleted()
+        {
+            const string id = "docs/1";
+            using (var store = GetDocumentStore())
+            {
+                using (var session = store.OpenSession())
+                {
+                    session.Store(new Document
+                    {
+                        Id = id,
+                        Text = "text-1"
+                    });
+                    session.SaveChanges();
+                }
+
+                using (var session = store.OpenSession())
+                {
+                    var doc = session.Load<Document>(id);
+                    Assert.NotNull(doc);
+
+                    session.Delete(doc);
+
+                    doc = new Document
+                    {
+                        Id = id,
+                        Text = "text-2"
+                    };
+                    
+                    session.Store(doc);
```

Comment:
I'm really nervous about something like that. It opens up the option that you'll be mislead by the code.
In your issue, you mentioned passing the id explicitly, which I think is a good compromise here.

Reply:
I'll see if I can make that work, thanks for the review.

Reply:
I do not think that this is the flow that we want to support.

Supporting following operations within a session is very dangerous
```
// load ID
// delete ID
// store ID
// saveChanges
```

What we should support is:
```
// load ID
// delete ID
// saveChanges
// store ID
// saveChanges
```

Reply:
The case I'm experiencing is adapting Quartz.NET's job store logic to pass connection (UoW) around template methods that handle various tasks, some might store, some might delete, the end result is then stored as transaction (session.SaveChanges()).

So it might not be obvious if composition caused load-delete-store combination, but SaveChanges should be the transaction.

Reply:
@ppekrol How you model this kind of templated process, where I need to support single transaction spanning composed methods? Composition methods are generic methods having generic type for passed connection/tx pair.

```
var uow = BeginTransactionalOperation(); // holder for ADO.NET connection+tx or Quartz.NET session
var job = LoadJob(trigger.JobId, uow);
var triggger = LoadJob(trigger.Id, uow);
MaybeDeleleteJobIfStale(job, uow);
MaybeAddJob(trigger, job, uow);
SomeOtherThing(trigger, job, uow);
uow.Commit();
```

Reply:
I've updated the PR to to only allow such overwrite if explicit id is provided, otherwise the goold old exception. I'll leave it to you to decide whether it makes sense or not - feel free to close and create and alternative solution. I don't get paid enough from OSS to fight the persistence layer 😄